src/Tests/dotnet-MsiInstallation.Tests/VirtualMachine.cs

```diff
@@ -0,0 +1,531 @@
+                    result.StdErr = results.Last().StdErr;
+                }
+                result.GroupedResults = results;
+
+                return result;
+            }
+            else
+            {
+                throw new NotImplementedException(action.Type.ToString());
+            }
+        }
+
+        public static string GetFileContentId(string path)
+        {
+            var info = new FileInfo(path);
+            return $"{info.LastWriteTimeUtc.Ticks}-{info.Length}";
+        }
+
+        public static string GetDirectoryContentId(string path)
+        {
+            StringBuilder sb = new StringBuilder();
+            var info = new DirectoryInfo(path);
+
+            void ProcessDirectory(DirectoryInfo dir, string relativeTo)
+            {
+                foreach (var file in dir.GetFiles())
+                {
+                    sb.AppendLine($"{Path.GetRelativePath(relativeTo, file.FullName)}:{file.LastWriteTimeUtc.Ticks}-{file.Length}");
+                }
+
+                foreach (var subDir in dir.GetDirectories())
+                {
+                    sb.AppendLine(subDir.FullName);
+                    ProcessDirectory(subDir, relativeTo);
+                }
+            }
+
+            ProcessDirectory(info, path);
+
+            return sb.ToString();
+        }
+
+        static void CopyDirectory(string sourcePath, string destPath)
+        {
+            if (!Directory.Exists(destPath))
+            {
+                Directory.CreateDirectory(destPath);
+            }
+
+            foreach (var dir in Directory.GetDirectories(sourcePath))
+            {
+                CopyDirectory(dir, Path.Combine(destPath, Path.GetFileName(dir)));
+            }
+
+            foreach (var file in Directory.GetFiles(sourcePath))
+            {
+                new FileInfo(file).CopyTo(Path.Combine(destPath, Path.GetFileName(file)), true);
+            }
+        }
+
+        string VMPathToSharePath(string vmPath)
+        {
+            var dirInfo = new DirectoryInfo(vmPath);
+
+            if (dirInfo.Root.FullName.Substring(1) != @":\")
+            {
+                throw new ArgumentException("Unrecognized directory root for: " + vmPath, nameof(vmPath));
+            }
+
+            string driveLetter = dirInfo.Root.FullName.Substring(0, 1);
+
+            string pathUnderDrive = dirInfo.FullName.Substring(3);
+
+            return $@"\\{VMControl.VMMachineName}\{driveLetter}$\{pathUnderDrive}";
+        }
+
+        //  Copilot wrote the entire body of this method
+        string SharePathToVMPath(string sharePath)
+        {
+            if (!sharePath.StartsWith($@"\\{VMControl.VMMachineName}\"))
+            {
+                throw new ArgumentException("Unrecognized share path: " + sharePath, nameof(sharePath));
+            }
+
+            string pathUnderDrive = sharePath.Substring($@"\\{VMControl.VMMachineName}\".Length);
+
+            string driveLetter = pathUnderDrive.Substring(0, 1);
```

Comment:
```suggestion
            var driveLetter = Path.GetPathRoot(new FileInfo(pathUnderDrive).FullName);
```

Reply:
I don't know whether `Path.GetPathRoot` actually does the right thing here (and a quick look at the docs didn't really clarify it for me), so I think it's more understandable with the Substring call.